handler/openid/flow_explicit_token.go

```diff
@@ -53,7 +53,7 @@ func (c *OpenIDConnectExplicitHandler) PopulateTokenEndpointResponse(ctx context
 		return errors.WithStack(fosite.ErrInvalidGrant.WithHint("The OAuth 2.0 Client is not allowed to use the authorization grant \"authorization_code\"."))
 	}
 
-	sess, ok := requester.GetSession().(Session)
+	sess, ok := authorize.GetSession().(Session)
```

Comment:
This is not correct and will lead to unexpected behavior. We're using `requester` here because we want the session that is being passed to this function, not the one from the store. Otherwise, things like the claims will be lost. You can find an example of that here: https://github.com/ory/hydra/blob/master/oauth2/handler.go#L661-L679

I know that the internals are quite difficult to grasp and the naming of the variables isn't the best, so it's quite difficult to track this down but I think the existing implementation is correct.

Also keep in mind that fosite went through OpenID Connect certification, and `at_hash` is pretty much a base requirement to get this passing. You can find a list of all the certification tests here: https://github.com/ory/hydra/tree/master/docs/certification

In conclusion, I believe that this patch would actually break behavior (which is also why existing tests had to be updated) and not fix a bug. I would double check your implementation of fosite to figure out why at_hash is not being set.

Because fosite is kind of tricky to consume I always recommend using Hydra, but I know that this might not be an option for you.

Reply:
Thanks I will check my implementation again too.

Reply:
Let me confirm one more thing please.
When calling IssueExplicitIDToken() [here](https://github.com/ory/fosite/pull/396/files#diff-b077d7414bc0c942d0f2c7f8176ef987R75) stored authorize requester `authorize` is used, instead of `requester`. 
Is this expected?
Can you please give me some hint of `when AccessTokenHash  is set to the session in that authorize requester` ?

Reply:
You are in fact right here, I had to look through several lines of code to understand what's happening here but fortunately I remembered (the code is quite old). When passing things down the stores, we always pass the requester along. This is because fosite expects the store to hydrate the session  like `json.Marshall`, so as an argument, not as a return value.

You can find use of that here:

1. https://github.com/ory/hydra/blob/master/oauth2/fosite_store_sql.go#L333
2. https://github.com/ory/hydra/blob/8dd9bb195e0c71c3123f0a93fc0f244050d0f7e2/oauth2/fosite_store_sql.go#L305
3. https://github.com/ory/hydra/blob/8dd9bb195e0c71c3123f0a93fc0f244050d0f7e2/oauth2/fosite_store_sql.go#L167-L173

Because of that, the session always points to the same memory - regardless of the request itself. Therefore, `at_hash` is set on the session and it does not matter if the request from the store or from the argument is passed to `IssueExplicitIDToken`.

I'm sorry that this is so complicated, the code is quite old and could need some refactoring. The problem with changing small things like that is that I have to run a full test suite against OIDC (which is very time consuming) to ensure that everything still works as expected. I'm not sure if we have `at_hash` tests in fosite for e2e.

Hope this helps

Reply:
Thank you for explaining. I will check!